### PR TITLE
Qchen/multi tenancy crd resource discovery

### PR DESCRIPTION
**Pull request by chenqianfzh**
This is a big PR involving multiple components. I could not split it into multiple PRs, as it will break the integration tests since these components are inter-related.  So I decide to write this length explanation, hoping it will help my reviewers gaining an easier experience.

### Resource discovery in K8s

Resource discovery means showing the group/version/kinds supported in the K8s cluster. When a CRD is created, the custom resource type will be displayed just like native resources.

There are two kubectl commands for resource discovery:
	**kubectl api-resources**
	**kubectl api-versions**

Besides, K8s supports the following APIServer urls to discover user-defined CRDs:
1. root url: **[api_server_IP]:[port]**
it displays the list of all the resource groups/groupVersions, including CRD
2. apis url: **[api_server_IP]:[port]/apis**
it displays the list of all the #core# and #CRD# resource groups/groupVersions 
3. resource group url: **[api_server_IP]:[port]/apis/[CRD_Group]**
it displayed the info about the CRD group
4. resource groupVersion url: **[api_server_IP]:[port]/apis/[CRD_Group]/[CRD_version]**
it displays the resource info in this CRD groupVersion

Note these urls are not only used by end users, but also by the components of K8s. 

### Multi-tenancy change for CRD resource discovery

With the code change in this PR, all the above commands/url formats now supports multi-tenancy.

In more details, the following commands will show the resource/groupVersion for the tenant specified in the context.
	**kubectl api-resources --context=user-context**
	**kubectl api-versions --context=user-context**

Also, the following tenant-aware resource discovery urls will be supported.
1. root url: **[api_server_IP]:[port]?tenant=###**
2. apis url: **[api_server_IP]:[port]/apis?tenant=###**
3. resource group url: **[api_server_IP]:[port]/apis/[CRD_Group]?tenant=###**
4. resource groupVersion url: **[api_server_IP]:[port]/apis/[CRD_Group]/[CRD_version]?tenant=###**

The new addition of "?tenant=###", in the resource discovery url, can be thought as the corresponding part of our addition of "/tenants/###/"  in a resource request (ex. /api/v1/tenants/t1/namespaces/ns1/pods/p1). 

If the tenant param is not given in the url, the system will fall back to userInfo in the request context to get the tenant info. It is similar to what short-path does in the resource request handling.

## About the PR

The handlers of every resource discovery url formats need to be changed. And the change of each handlers might involve some other components. What makes things even worse is that all these components are inter-related. The integration test fails if I send out a PR including just one of the changes. Thus I have this jumbo PR. 

I hope the PR review experience would be a bit easier by splitting the code into several commits based on their functions. The key commits in this PR:
1. commit titled **"multi-tenancy resource discovery at root url"**

this commit makes the handler of "root url" ([api_server_IP]:[port]) support multi-tenancy.

The exising design register all the group/groupVersion discovery urls. I change it to register the tenant-aware urls for CRDs, which inlcudes the "?tenant=" part. When a request comes in, it only returns the the url has the matching tenant. Also the "?tenant=" part will be removed before return so a regular tenant will have the same experience as in K8s.

2. commit titled **"multi-tenancy CRD group/groupversion resource discovery"**

this commit makes the handler of [api_server_IP]:[port]/apis/[CRD_Group] and [api_server_IP]:[port]/apis/[CRD_Group]/[CRD_version] support multi-tenancy.

The existing design keeps a map, each group/groupVersion maps to a resource discovery handler. I changed the map to a double-layer map, the first-layer key is the tenant, and the second layer key is the group/groupVersion, so differennt tenants can have different resource handlers even they have the same group/groupVersion.

3. commit titled **"make APIGroup and APIs resource discovery support multi-tenancy"**

this commit makes the handler of [api_server_IP]:[port]/apis support multi-tenancy.
This handler belongs to kube-aggregator (a design I am not quite sure why). it converts each resource, either native or custom, into an APIService. The multi-tenancy changes happened to several related components:
	a. CRD registration controller: this controller notifies APIService auto register controller when it detects a new CRD
	b. The APIService auto register controller: this controller register each new APIService
	c. Surely, the handlers themselves.

Obviously, the APIService resource object needs to be changed to tenant-scoped (it was cluster-scoped). So there are two auxiliary commits about APIService scope change for this handler change:
	a. commit titled **"change the scope of APIService from cluster-scope to tenant-scope"**
        b. commit titled **"auto updates after scope change of APIService"**

The other auxiliary commits include:
1. "test changes"
2. "common changes for multiple resource discovery handlers"

## Test In My Dev Box

I created two tenants, qian and peng. Also two corresponding user contexts, qian-admin-context and peng-admin-context. I create a CRD under tenant qian first, and check the result. Then I create the same CRD under tenant peng, and does the check. Finally, I removed the CRD under tenant peng, and check again.

For each check, I run the command "kubectl api-resources" and "kubectl api-versions" for each tenant. I also run each tenant-aware resource discovery url for each tenant. The kubectl commands covers the scenario that tenant is not specified as a param in the url and the system get the tenant info from request context.

### First, I created CRD foo.samplecontroller.k8s.io under the tenant of qian
```
prompt$ kubectl --context=qian-admin-context apply -f /home/qianchen/gopath/qianc/arktos/staging/src/k8s.io/sample-controller/artifacts/examples/crd.yaml
customresourcedefinition.apiextensions.k8s.io/foos.samplecontroller.k8s.io created

prompt$ kubectl --context=qian-admin-context apply -f /home/qianchen/gopath/qianc/arktos/staging/src/k8s.io/sample-controller/artifacts/examples/example-foo.yaml
foo.samplecontroller.k8s.io/example-foo created
prompt$ kubectl api-resources --context=qian-admin-context | grep sample
foos                                      samplecontroller.k8s.io     true       true         Foo

prompt$ kubectl api-versions --context=qian-admin-context | grep sample
samplecontroller.k8s.io/v1alpha1

prompt$ kubectl get foos  --context=qian-admin-context
NAME          AGE
example-foo   16s


prompt$ kubectl api-versions --context=peng-admin-context | grep sample

prompt$ kubectl api-resources --context=peng-admin-context | grep sample

prompt$ kubectl get foos  --context=peng-admin-context
error: the server doesn't have a resource type "foos"
```
The test results for the tenant-ware resource discovery urls

**http://localhost:8080/?tenant=qian**

....
47	"/apis/samplecontroller.k8s.io/v1alpha1"
48	"/apis/samplecontroller.k8s.io"
...

**http://localhost:8080/apis/?tenant=qian**
18	
	name	"samplecontroller.k8s.io"
	versions	
		0	
			groupVersion	"samplecontroller.k8s.io/v1alpha1"
			version	"v1alpha1"
	preferredVersion	
		groupVersion	"samplecontroller.k8s.io/v1alpha1"
		version	"v1alpha1"

**http://localhost:8080/apis/samplecontroller.k8s.io?tenant=qian**	
kind	"APIGroup"
apiVersion	"v1"
name	"samplecontroller.k8s.io"
versions	
	0	
		groupVersion	"samplecontroller.k8s.io/v1alpha1"
		version	"v1alpha1"
preferredVersion	
	groupVersion	"samplecontroller.k8s.io/v1alpha1"
	version	"v1alpha1"

**http://localhost:8080/apis/samplecontroller.k8s.io/v1alpha1?tenant=qian**
kind	"APIResourceList"
apiVersion	"v1"
groupVersion	"samplecontroller.k8s.io/v1alpha1"
resources	
	0	
		name	"foos"
		singularName	"foo"
		namespaced	true
		kind	"Foo"
		verbs	
			0	"delete"
			1	"deletecollection"
			2	"get"
			3	"list"
			4	"patch"
			5	"create"
			6	"update"
			7	"watch"
		storageVersionHash	"H8mNWb07qhE="
		tenanted	true

http://localhost:8080/?tenant=peng
Nothing about samplecontroller is displayed

http://localhost:8080/apis/?tenant=peng
Nothing about samplecontroller is displayed

http://localhost:8080/apis/samplecontroller.k8s.io?tenant=peng
404 page not found

http://localhost:8080/apis/samplecontroller.k8s.io/v1alpha1?tenant=peng
404 page not found

### Next, I created the same CRD under tenant peng
```
prompt$ kubectl --context=peng-admin-context apply -f /home/qianchen/gopath/qianc/arktos/staging/src/k8s.io/sample-controller/artifacts/examples/crd.yaml
customresourcedefinition.apiextensions.k8s.io/foos.samplecontroller.k8s.io created

prompt$ kubectl --context=peng-admin-context apply -f /home/qianchen/gopath/qianc/arktos/staging/src/k8s.io/sample-controller/artifacts/examples/example-foo.yaml
foo.samplecontroller.k8s.io/example-foo created

prompt$ kubectl api-resources --context=qian-admin-context | grep sample
foos                                      samplecontroller.k8s.io     true       true         Foo

prompt$ kubectl api-versions --context=qian-admin-context | grep sample
samplecontroller.k8s.io/v1alpha1

prompt$ kubectl get foos  --context=qian-admin-context
NAME          AGE
example-foo   52m

prompt$ kubectl api-versions --context=peng-admin-context | grep sample
samplecontroller.k8s.io/v1alpha1

prompt$ kubectl api-resources --context=peng-admin-context | grep sample
foos                                      samplecontroller.k8s.io     true       true         Foo

prompt$ kubectl get foos  --context=peng-admin-context
NAME          AGE
example-foo   4s
```

I also test the tenant-aware urls of all the formats for tenant qian and peng. All the urls show correct results. The lengthy results are skipped for a better reading experience.

### Next, I removed the CRD under the tenant peng
```
prompt$ kubectl --context=peng-admin-context delete -f /home/qianchen/gopath/qianc/arktos/staging/src/k8s.io/sample-controller/artifacts/examples/crd.yaml
customresourcedefinition.apiextensions.k8s.io "foos.samplecontroller.k8s.io" deleted

prompt$ kubectl api-resources --context=qian-admin-context | grep sample
foos                                      samplecontroller.k8s.io     true       true         Foo

prompt$ kubectl api-versions --context=qian-admin-context | grep sample
samplecontroller.k8s.io/v1alpha1

prompt$ kubectl get foos  --context=qian-admin-context
NAME          AGE
example-foo   58m

prompt$ kubectl api-versions --context=peng-admin-context | grep sample

prompt$ kubectl api-resources --context=peng-admin-context | grep sample

prompt$ kubectl get foos  --context=peng-admin-context
error: the server doesn't have a resource type "foos"
```
The test for the tenant-aware urls of all the formats for tenant qian and peng give the same result as what we see when the CRD was not created under tenant peng.

### CODE REVIEW

**Comment by pdgetrf** [comment]
http://localhost:8080/?tenant=qian
http://localhost:8080/?tenant=peng

do these allow one tenant to discover the resource of another tenant? If yes, is this behavior allowed?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery_controller.go [hunk 1]
```diff
@@ -74,24 +80,27 @@ func NewDiscoveryController(crdInformer informers.CustomResourceDefinitionInform
...
+func (c *DiscoveryController) sync(gvt GroupVersionTenant) error {
 
 	apiVersionsForDiscovery := []metav1.GroupVersionForDiscovery{}
 	apiResourcesForDiscovery := []metav1.APIResource{}
 	versionsForDiscoveryMap := map[metav1.GroupVersion]bool{}
 
-	crds, err := c.crdLister.List(labels.Everything())
+	groupVersion := schema.GroupVersion{Group: gvt.group, Version: gvt.version}
+
+	crds, err := c.crdLister.CustomResourceDefinitionsWithMultiTenancy(gvt.tenant).List(labels.Everything())
```
Comment (pdgetrf) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery_controller.go:91]: nit: CustomResourceDefinitionsWithMultiTenancy --> CRDWithMultiTenancy

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery_controller.go [hunk 2]
```diff
@@ -112,7 +122,8 @@ func (c *DiscoveryController) sync(version schema.GroupVersion) error {
 					Version:      v.Name,
 				})
 			}
-			if v.Name == version.Version {
+
+			if v.Name == gvt.version {
```
Comment (pdgetrf) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery_controller.go:126]: version vs Version? Is there any reason why it was Capitalized V?

File: staging/src/k8s.io/apiserver/pkg/server/routes/index.go [hunk 3]
```diff
@@ -65,5 +69,18 @@ type IndexLister struct {
 
 // ServeHTTP serves the available paths.
 func (i IndexLister) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```
Comment (pdgetrf) [inline staging/src/k8s.io/apiserver/pkg/server/routes/index.go:71]: add a unit test for this ServeHTTP if there isn't one already

File: staging/src/k8s.io/apiserver/pkg/server/routes/index.go [hunk 4]
```diff
@@ -65,5 +69,18 @@ type IndexLister struct {
...
+	userTenant, _ := apifilters.GetTenantFromQueryThenContext(r)
+	tenantedPathMark := "?tenant="
+	tenantPathSuffix := tenantedPathMark + userTenant
+	rootPaths := []string{}
+	for _, path := range i.PathProvider.ListedPaths() {
+		if !strings.Contains(path, tenantedPathMark) {
+			rootPaths = append(rootPaths, path)
+		} else {
+			if strings.HasSuffix(path, tenantPathSuffix) {
+				rootPaths = append(rootPaths, strings.TrimSuffix(path, tenantPathSuffix))
```
Comment (pdgetrf) [inline staging/src/k8s.io/apiserver/pkg/server/routes/index.go:81]: add a unit test for this ServeHTTP if there isn't one already

File: staging/src/k8s.io/client-go/discovery/discovery_client.go [hunk 5]
```diff
@@ -200,7 +208,13 @@ func (d *DiscoveryClient) ServerResourcesForGroupVersion(groupVersion string) (r
 	resources = &metav1.APIResourceList{
 		GroupVersion: groupVersion,
 	}
-	err = d.RESTClient().Get().AbsPath(url.String()).Do().Into(resources)
+
+	if !tenantedFormat {
+		err = d.RESTClient().Get().AbsPath(url.String()).Do().Into(resources)
+	} else {
+		err = d.RESTClient().Get().AbsPath(url.String()).Param("tenant", parts[0]).Do().Into(resources)
```
Comment (pdgetrf) [inline staging/src/k8s.io/client-go/discovery/discovery_client.go:215]: string constant "tenant" is also used at another place. Is it possible to put this into a shareable constant?

File: staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/helpers.go [hunk 6]
```diff
@@ -86,9 +87,16 @@ func (s ByVersionPriority) Less(i, j int) bool {
 
 // APIServiceNameToGroupVersion returns the GroupVersion for a given apiServiceName.  The name
 // must be valid, but any object you get back from an informer will be valid.
-func APIServiceNameToGroupVersion(apiServiceName string) schema.GroupVersion {
+func APIServiceNameToGroupVersionTenant(tenantedApiServiceName string) (string, string, schema.GroupVersion) {
```
Comment (pdgetrf) [inline staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/helpers.go:90]: likely need a unit test if there's not one yet

File: staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/validation/validation.go [hunk 7]
```diff
@@ -33,7 +33,7 @@ import (
 func ValidateAPIService(apiService *apiregistration.APIService) field.ErrorList {
 	requiredName := apiService.Spec.Version + "." + apiService.Spec.Group
 
-	allErrs := validation.ValidateObjectMeta(&apiService.ObjectMeta, false, false,
+	allErrs := validation.ValidateObjectMeta(&apiService.ObjectMeta, true, false,
```
Comment (pdgetrf) [inline staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/validation/validation.go:36]: why the reversed boolean?

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go [hunk 8]
```diff
@@ -287,31 +295,39 @@ func (s *APIAggregator) AddAPIService(apiService *apiregistration.APIService) er
...
 
-	// TODO unregister group level discovery when there are no more versions for the group
-	// We don't need this right away because the handler properly delegates when no versions are present
+	groupPath := "/apis/" + version.Group
+	s.GenericAPIServer.Handler.NonGoRestfulMux.Unregister(groupPath + suffix)
+	s.GenericAPIServer.Handler.NonGoRestfulMux.Unregister(groupPath + suffix + "/")
+	delete(s.handledGroups, tenantedGroup)
 }
 
 // DefaultAPIResourceConfigSource returns default configuration for an APIResource.
```
Comment (pdgetrf) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:333]: long PR. going out for a smoke. back soon.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_apis.go [hunk 9]
```diff
@@ -58,18 +60,50 @@ var discoveryGroup = metav1.APIGroup{
...
 }
 
+// getApiServices return the list of APIService for the tenant in the request,
+// including the native APIServices, which is under the system tenant
+// and the APIServices corresponding to the CRDs created by the tenant in the request
+func getApiServices(lister listers.APIServiceLister, req *http.Request) ([]*apiregistrationapi.APIService, error) {
+	// Note: there is an issue here, the CRD created by the system tenant will be visible to all the tenants.
+	// TODO: this issue will be addressed in the following work of CRD sharing
+	// tracking issue: https://github.com/futurewei-cloud/arktos/issues/292
+	systemApiServices, err := lister.APIServicesWithMultiTenancy(metav1.TenantSystem).List(labels.Everything())
```
Comment (pdgetrf) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_apis.go:70]: why is TenantSystem used here?

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiservice_controller.go [hunk 10]
```diff
@@ -37,7 +38,7 @@ import (
 // APIHandlerManager defines the behaviour that an API handler should have.
 type APIHandlerManager interface {
 	AddAPIService(apiService *apiregistration.APIService) error
-	RemoveAPIService(apiServiceName string)
+	RemoveAPIService(tenantedApiServiceName string)
```
Comment (pdgetrf) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiservice_controller.go:41]: why the renaming? is this needed?

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_apis_test.go [hunk 11]
```diff
@@ -292,7 +295,7 @@ func TestAPIs(t *testing.T) {
 		server := httptest.NewServer(handler)
 		defer server.Close()
 
-		resp, err := http.Get(server.URL + "/apis")
+		resp, err := http.Get(server.URL + "/apis" + "?tenant=" + testTenant)
```
Comment (pdgetrf) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_apis_test.go:298]: can this "?tenant=" + tenant be put into a function to be shared by code and tests?

File: staging/src/k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/apiregistration/v1/apiservice.go [hunk 12]
```diff
@@ -57,20 +58,27 @@ type APIServiceInterface interface {
 type aPIServices struct {
 	client  rest.Interface
 	clients []rest.Interface
+	te      string
```
Comment (pdgetrf) [inline staging/src/k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/apiregistration/v1/apiservice.go:61]: te --> tenant

File: staging/src/k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/apiregistration/v1/apiservice.go [hunk 13]
```diff
@@ -57,20 +58,27 @@ type APIServiceInterface interface {
...
 // newAPIServices returns a APIServices
 func newAPIServices(c *ApiregistrationV1Client) *aPIServices {
+	return newAPIServicesWithMultiTenancy(c, "system")
+}
+
+func newAPIServicesWithMultiTenancy(c *ApiregistrationV1Client, tenant string) *aPIServices {
 	return &aPIServices{
 		client:  c.RESTClient(),
 		clients: c.RESTClients(),
+		te:      tenant,
```
Comment (pdgetrf) [inline staging/src/k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/apiregistration/v1/apiservice.go:73]: same here te --> tenant

File: staging/src/k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/apiregistration/v1/apiservice.go [hunk 14]
```diff
@@ -157,7 +168,13 @@ func (c *aPIServices) Watch(opts metav1.ListOptions) watch.AggregatedWatchInterf
 func (c *aPIServices) Create(aPIService *v1.APIService) (result *v1.APIService, err error) {
 	result = &v1.APIService{}
 
+	objectTenant := aPIService.ObjectMeta.Tenant
+	if objectTenant == "" {
```
Comment (pdgetrf) [inline staging/src/k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/apiregistration/v1/apiservice.go:172]: could objectTenant ever be empty?

Comment (pdgetrf) [reply staging/src/k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/apiregistration/v1/apiservice.go:172]: or this is generated code

File: staging/src/k8s.io/kube-aggregator/pkg/client/informers/internalversion/apiregistration/internalversion/apiservice.go [hunk 15]
```diff
@@ -51,23 +52,31 @@ func NewAPIServiceInformer(client internalclientset.Interface, resyncPeriod time
...
 
+func NewAPIServiceInformerWithMultiTenancy(client internalclientset.Interface, resyncPeriod time.Duration, indexers cache.Indexers, tenant string) cache.SharedIndexInformer {
+	return NewFilteredAPIServiceInformerWithMultiTenancy(client, resyncPeriod, indexers, nil, tenant)
+}
+
 // NewFilteredAPIServiceInformer constructs a new informer for APIService type.
 // Always prefer using an informer factory to get a shared informer instead of getting an independent
 // one. This reduces memory footprint and number of connections to the server.
 func NewFilteredAPIServiceInformer(client internalclientset.Interface, resyncPeriod time.Duration, indexers cache.Indexers, tweakListOptions internalinterfaces.TweakListOptionsFunc) cache.SharedIndexInformer {
+	return NewFilteredAPIServiceInformerWithMultiTenancy(client, resyncPeriod, indexers, tweakListOptions, "system")
```
Comment (pdgetrf) [inline staging/src/k8s.io/kube-aggregator/pkg/client/informers/internalversion/apiregistration/internalversion/apiservice.go:63]: what is "system"?

Comment (pdgetrf) [reply staging/src/k8s.io/kube-aggregator/pkg/client/informers/internalversion/apiregistration/internalversion/apiservice.go:63]: also consider add a short comment about why a const tenant is used here to make sure it's on purpose.

File: staging/src/k8s.io/kube-aggregator/pkg/client/listers/apiregistration/internalversion/apiservice.go [hunk 16]
```diff
@@ -63,3 +67,52 @@ func (s *aPIServiceLister) Get(name string) (*apiregistration.APIService, error)
...
+	err = cache.ListAllByTenant(s.indexer, s.tenant, selector, func(m interface{}) {
+		ret = append(ret, m.(*apiregistration.APIService))
+	})
+	return ret, err
+}
+
+// Get retrieves the APIService from the indexer for a given tenant and name.
+func (s aPIServiceTenantLister) Get(name string) (*apiregistration.APIService, error) {
+	key := s.tenant + "/" + name
+	if s.tenant == "system" {
```
Comment (pdgetrf) [inline staging/src/k8s.io/kube-aggregator/pkg/client/listers/apiregistration/internalversion/apiservice.go:107]: same with "system" here.

File: test/integration/etcd/data.go [hunk 17]
```diff
@@ -580,7 +580,8 @@ func GetCustomResourceDefinitionData() []*apiextensionsv1beta1.CustomResourceDef
 		// namespaced with legacy version field
 		{
 			ObjectMeta: metav1.ObjectMeta{
-				Name: "foos.cr.bar.com",
+				Name:   "foos.cr.bar.com",
+				Tenant: metav1.TenantSystem,
```
Comment (pdgetrf) [inline test/integration/etcd/data.go:584]: why the system tenant here?

File: staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/validation/validation.go (hunk 7 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/validation/validation.go:36]: It indicates the scope is changed from cluster-scoped to tenant-scoped. True means that the validator will check to make sure that "Tenant" in ObjectMeta is not empty.

**Comment by chenqianfzh** [comment]
> http://localhost:8080/?tenant=qian
> http://localhost:8080/?tenant=peng
> 
> do these allow one tenant to discover the resource of another tenant? If yes, is this behavior allowed?

Yes, it will allow a tenant to discover resources of another tenant. Especially, we need to allow the system tenant to discover those of all the other regular tenants. 

The work in this PR is about resource discovery. How to constrain the discovery is the job of authorization.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery_controller.go (hunk 1 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery_controller.go:91]: Cannot do it. :-(

CustomResourceDefinitionsWithMultiTenancy is a name auto-generated by client-go.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiservice_controller.go (hunk 10 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiservice_controller.go:41]: it is not necessary, but it helps code reading. 

apiServiceName is APIService.Name
tenantedApiServiceName is APIService.Tenant + "/" + APIService.Name.

variable name "apiServiceName" is used everywhere. Renaming helps identifying the difference.

File: test/integration/etcd/data.go (hunk 17 above)
Comment (chenqianfzh) [reply test/integration/etcd/data.go:584]: data.go is the test cases for backward-compatibility, where the tenant is "system". There is another file called multi_tenancy_data.go, where the non-system-tenant test cases are listed.

I added the Tenant here to call it out the difference with the test cases in multi_tenancy_data.go, though in real life the empty Tenant in ObjectMeta will be set by the short-path logic.

File: staging/src/k8s.io/kube-aggregator/pkg/client/informers/internalversion/apiregistration/internalversion/apiservice.go (hunk 15 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/kube-aggregator/pkg/client/informers/internalversion/apiregistration/internalversion/apiservice.go:63]: This is a file auto-generated by codegen. Codegen is not smart enough to import the dependency needed for a constant variable (it is good when importing dependencies for  types/packages/funcs). So I don't have a way to include the dependency of package "metav1" in the file generated.

File: staging/src/k8s.io/kube-aggregator/pkg/client/listers/apiregistration/internalversion/apiservice.go (hunk 16 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/kube-aggregator/pkg/client/listers/apiregistration/internalversion/apiservice.go:107]: Same explanation as above.

File: staging/src/k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/apiregistration/v1/apiservice.go (hunk 14 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/apiregistration/v1/apiservice.go:172]: Yes, it is generated code.

File: staging/src/k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/apiregistration/v1/apiservice.go (hunk 12 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/apiregistration/v1/apiservice.go:61]: Agree, te is a bad naming. However, this is auto-generated code. I need to change to the code generator template to make the change, which will affect hundreds of files. 

Let me make this refactoring in another PR.

File: staging/src/k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/apiregistration/v1/apiservice.go (hunk 13 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/kube-aggregator/pkg/client/clientset_generated/clientset/typed/apiregistration/v1/apiservice.go:73]: Same as above, will refactor it in another PR.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery_controller.go (hunk 2 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery_controller.go:126]: The upper-case Version is from the type schema.GroupVersion, which is a type used in a lot of locations. So its members are public, in upper cases.

gvt is of type GroupVersionTenant, which is a type I define to use in this controller only. Therefore I made the members private, in lower cases.

File: staging/src/k8s.io/client-go/discovery/discovery_client.go (hunk 5 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/client-go/discovery/discovery_client.go:215]: Fixed.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_apis_test.go (hunk 11 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_apis_test.go:298]: Updated as suggested

File: staging/src/k8s.io/apiserver/pkg/server/routes/index.go (hunk 4 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/apiserver/pkg/server/routes/index.go:81]: Added.

File: staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/helpers.go (hunk 6 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/kube-aggregator/pkg/apis/apiregistration/helpers.go:90]: Yep, added.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/tenantinfo.go [hunk 18]
```diff
@@ -55,11 +56,65 @@ func SetShortPathRequestTenant(req *http.Request) (*http.Request, error) {
...
+	u, _ := url.Parse(urlString)
+	tenantValues, ok := u.Query()[TenantParam]
+	if ok {
+		return tenantValues[0]
+	}
+
+	return ""
+}
+
+func GetTenantFromQuery(req *http.Request) string {
```
Comment (pdgetrf) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/tenantinfo.go:94]: nit: is this one line function needed?

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go [hunk 19]
```diff
@@ -247,6 +252,10 @@ func (s *APIAggregator) AddAPIService(apiService *apiregistration.APIService) er
 	}
 
 	proxyPath := "/apis/" + apiService.Spec.Group + "/" + apiService.Spec.Version
+	suffix := ""
```
Comment (pdgetrf) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:255]: why the empty suffix?

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go [hunk 20]
```diff
@@ -287,31 +296,39 @@ func (s *APIAggregator) AddAPIService(apiService *apiregistration.APIService) er
...
 // It's a slow moving API, so it's ok to run the controller on a single thread.
-func (s *APIAggregator) RemoveAPIService(apiServiceName string) {
-	version := apiregistration.APIServiceNameToGroupVersion(apiServiceName)
+func (s *APIAggregator) RemoveAPIService(tenantedApiServiceName string) {
+	tenant, apiServiceName, version := apiregistration.APIServiceNameToGroupVersionTenant(tenantedApiServiceName)
+	tenantedGroup := tenant + "/" + version.Group
+	tenatedApiServiceName := tenant + "/" + apiServiceName
 
 	proxyPath := "/apis/" + version.Group + "/" + version.Version
+	suffix := ""
```
Comment (pdgetrf) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:313]: also here.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/tenantinfo.go (hunk 18 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/tenantinfo.go:94]: You are right. Removed this func.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go (hunk 19 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:255]: Ah, a "covfefe"-style mistake. Thanks for catching this.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go (hunk 20 above)
Comment (chenqianfzh) [reply staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go]: fixed.

**Pull request merged**